rust/src/stats.rs

```diff
@@ -0,0 +1,76 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::format_one_arg;
+use reqwest::header::CONTENT_TYPE;
+use reqwest::header::USER_AGENT;
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use std::sync::mpsc::Sender;
+use std::time::Duration;
+
+const PLAUSIBLE_URL: &str = "https://plausible.io/api/event";
+const SM_USER_AGENT: &str = "Selenium Manager {}";
+const APP_JSON: &str = "application/json";
+const PAGE_VIEW: &str = "pageview";
+const SELENIUM_DOMAIN: &str = "manager.selenium.dev";
+const SM_STATS_URL: &str = "https://{}/sm-stats";
+const REQUEST_TIMEOUT_SEC: u64 = 3;
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct Data {
+    pub name: String,
+    pub url: String,
+    pub domain: String,
+    pub props: Props,
+}
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct Props {
+    pub browser: String,
+    pub browser_version: String,
+    pub os: String,
+    pub arch: String,
+    pub lang: String,
+    pub selenium_version: String,
```

Comment:
I know we did not mention this before, but... should we add the driver version? What do you think @bonigarcia?

Reply:
I suppose it is not very relevant since most users would use the browser version (`browserVersion` set through options) but not the driver version.

BTW, the browser version sent to plausible is that value, i.e., what the user specifies through options (e.g., `"beta"`, `"120"`, `""`) and not the discovered browser version (e.g., "120.0.6099.71", etc.). Is that what we want, right?

Reply:
I think it makes sense to avoid the driver version then.

Regarding the browser version... I think we should collect both. Does `requested_browser_version` sound good?

Reply:
It can be done (gather both `requested_browser_version` and `browser_version`). But if we do that, we must send the stats after executing the driver/browser management. In other words, the call to `.stats()` is done after `.setup()`.

In that case, and since after the `.setup()` SM is almost finished (it only needs to log the output), we should wait for the `.stats()` to be finished (since now it is executed in a separate thread). Alternatively, we should execute both `.stats()` and `.setup()` in the same thread (as it was done in the original PR). In that case, the call to Plausible will include some ms overhead (as reported in the description of this PR).

Do we want to do that?

Reply:
That is a very good point. 

This also means that, with the current implementation, most of the time the `browser_version` value is empty because users normally don't specify a browser version.

To avoid loosing data, what if we execute `.stats()` in the same thread but shorten the read timeout to 1 second? In theory, we should not need to wait for a response. 

One last thing I noticed is that executing `.stats()` is not guarded by `selenium_manager.is_offline()`. I mean, if the user wants an offline execution, we should not send stats either.

Reply:
Can't we create the second process at the end of the setup() and not worry about the timeout?

Reply:
The problem is that the binary will exit and we might not send the POST completely.

Reply:
ah, so we're just taking advantage of the fact that setup() takes longer than stats() with the current implementation?

Reply:
I would say so.

Reply:
I actually think it is ok not to have discovered version. All that tells us is what the user happened to have installed on their machine? I don't think it is worth the effort to capture it...

Reply:
Why not? I think it is not a big effort, we are just trying to be careful to avoid blocking the user. I think the data is worth to have.

Reply:
My point is that if it adds a second to each test it's not worth the tradeoff. I'm more interested in what people request than what they have installed. Most likely what we'll see installed is the latest version of chrome because it auto updates.

Reply:
Actually, this is a good point. Enabling Selenium Manager on Grid is going to look different than using it in client. Can we differentiate when grid is requesting and when the Java client is rquesting?

Reply:
It does not add a second to each test, it should be a few ms as Boni pointed out. Worst case, it would add 1 second. We could even shorten that to 500ms.

Reply:
Can we just fire the event and not wait for the response since we aren't going to do anything with it, anyway?

Reply:
I like the current implementation because it does not impact the overall performance. Moving the `.stats()` call to the end of the SM logic will affect some ms each time SM is executed. It is worth mentioning that in my tests (i.e., using the GHA infrastructure), this overhead is an average of 32 ms. But for any user, this figure will depend on the underlying network latency.

I tried to implement a *fire and forget* approach for the POST requests to Plausible in the Rust logic. But if the client doesn't wait for the response, Plausible does not count it. So, for gathering the discovered browser version (I can also include the found driver version for the same prize), I'm afraid we need to wait for the response, adding some MS overhead per test.

Reply:
I have just added a commit to check the offline mode.

Please let me know your final decision about the current discussion. There are two alternatives:

a) Move the call to `stats()` at the end of SM logic and add the discovered browser version (and found driver version if you want) to the data sent to Plausible. This call should be blocking, and so, it will add a little overhead.
b) Leave it in the current way, in which the browser version is the one reported by the user and `stats()` is done at the beginning (concurrently).

Reply:
I think we could go with option a) and iterate. Maybe we do not need a perfect solution now, just something that helps us understand usage.

Reply:
My vote is to keep existing implementation and just get the browser-version passed by the user because that's the least impact.

Reply:
"usage" comes from what the user passes in. Knowing what a user has on their system seems extra. I think we should start with the least-impact and add to it if we think we need more details.

Reply:
🤦 

I meant option `b)`, so as it currently is.

![](https://media.giphy.com/media/ii7CbaW46Bi02XCQr0/giphy.gif)

Reply:
My vote, if I have one, is b. For me, it is better if this feature does not impacting the overall performance, even if it is a few ms.

Reply:
@bonigarcia,  of course you have a say!